### PR TITLE
Proposed error-reporting ADR

### PR DESCRIPTION
**Pull request by devjgm**
The proposed error-reporting ARD states that we will switch to using `StatusOr<T>` for all error reporting instead of throwing exceptions. This has been discussed offline; this ARD is an attempt to capture the decision.

NOTE: This is the first ARD that I've written. I'm not sure how much information should go into the doc. If I need to add more, I'm happy to do so. Just let me know.

[skip ci]

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1804)

### CODE REVIEW

File: doc/adr/2019-01-04-error-reporting-with-statusor.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+**Title**: Functions should report errors using `StatusOr<T>` *instead of*
+throwing exceptions.
+
+**Status**: proposed
+
+**Context**: We know there will be users of these C++ libraries who want to use
+C++ exceptions as well as those who are not able to. Our C++ libraries must
+work for all of our users, regardless of their ability to use exceptions.
+
+**Decision**: All of our APIs will report errors to callers by returning a
```
Comment (coryan) [inline doc/adr/2019-01-04-error-reporting-with-statusor.md]: "All of our APIs" may be tricky: what about things that return `void`?  Are we going to return `StatusOr<void>` or `Status`?  What about returning an object derived from `std::istream` (or `std::ostream`) which already have a mechanism to report errors (the `badbit`)?

Consider starting with "None of our APIs will throw exceptions..." (modulo platforms that raise `std::bad_alloc` but I digress), and then saying "Our APIs will return `StatusOr<T>` <placeholder> to indicate an error. The only exception would be APIs that have existing error reporting mechanisms, e.g., `std::ostream`"

In `<placeholder>` you may want to say "(except APIs that return `void` in which case the API will return `Status`)" or something like that.

File: doc/adr/2019-01-04-error-reporting-with-statusor.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+**Title**: Functions should report errors using `StatusOr<T>` *instead of*
+throwing exceptions.
+
+**Status**: proposed
```
[the remark this answers is not in the document]
Comment (devjgm) [reply doc/adr/2019-01-04-error-reporting-with-statusor.md]: Ack. I changed it now. Obviously, we won't approve this change if it's not accepted :-)

File: doc/adr/2019-01-04-error-reporting-with-statusor.md (hunk 1 above)
Comment (coryan) [reply doc/adr/2019-01-04-error-reporting-with-statusor.md]: I forgot to mention, if we return `Status` instead of `StatusOr<void>` there is no `.value()` to raise an exception from.

File: doc/adr/2019-01-04-error-reporting-with-statusor.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+A downside of this decision is that our APIs will not be natural or idiomatic
+for the [50+%][survey-link] of users who might prefer exceptions for error
+reporting.
+
+Changing existing APIs from throwing exceptions to returning `StatusOr<T>` is a
+breaking change. As of this writing (Jan 2019), this project has a [Google
+Cloud Storage][gcs-link] component that is at the Alpha quality level, and a
+[Google Cloud Bigtable][bigtable-link] that is already at the Beta quality
+level. Since neither of these are at the GA quality level, breaking changes are
```
[the remark this answers is not in the document]
Comment (devjgm) [reply doc/adr/2019-01-04-error-reporting-with-statusor.md]: I like your wording better. Changed.

File: doc/adr/2019-01-04-error-reporting-with-statusor.md (hunk 1 above)
Comment (devjgm) [reply doc/adr/2019-01-04-error-reporting-with-statusor.md]: All good. Incorporated.

File: doc/adr/2019-01-04-error-reporting-with-statusor.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+C++ exceptions as well as those who are not able to. Our C++ libraries must
+work for all of our users, regardless of their ability to use exceptions.
+
+**Decision**: None of our APIs will throw exceptions to indicate errors.
+Instead, our APIs will typically report errors to callers by returning a
+`Status` or `StatusOr<T>` object, unless the library we're using has another
+non-throwing way to report errors (e.g., [badbit][badbit-link] in the standard
+I/O library).
+
+**Decision**: All of our APIs will report errors to callers by returning a
```
Comment (coryan) [inline doc/adr/2019-01-04-error-reporting-with-statusor.md]: Did you mean to remove the second "Decision"?

**Comment by coryan** [comment]
Remember `[skip ci]` is your friend.

File: doc/cpp-style-guide.md [hunk 5]
```diff
@@ -392,36 +387,6 @@ other classes solely through their public members.
...
-- Failing to contact the server should raise an exception.
-- Receiving an invalid response from the server should raise an exception.
-- Receiving a permanent error status from the server should raise an exception.
-- Being unable to find an object, or row, or file in a remote server should
-  not raise an exception.
-
-As we said, there are no hard and fast rules, consult your colleagues and be
-ready to change your mind.
-
-This is substantially different from the corresponding
```
Comment (coryan) [inline doc/cpp-style-guide.md]: I think you still want a section that says:

----
#### Exceptions

We do not use C++ exceptions.

[GSG section](https://google.github.io/styleguide/cppguide.html#Exceptions)
----

We may want to completely remove this document at this point, since there are no major differences now, but that should be a separate PR.

File: doc/adr/2019-01-04-error-reporting-with-statusor.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+their own code.
+
+A downside of this decision is that our APIs will not be natural or idiomatic
+for the [50+%][survey-link] of users who might prefer exceptions for error
+reporting.
+
+Changing existing APIs from throwing exceptions to returning `StatusOr<T>` is a
+breaking change. As of this writing (Jan 2019), this project has a [Google
+Cloud Storage][gcs-link] component that is at the Alpha quality level, and a
+[Google Cloud Bigtable][bigtable-link] that is already at the Beta quality
```
Comment (devbww) [inline doc/adr/2019-01-04-error-reporting-with-statusor.md]: s/that/component that/
